core/core.go

```diff
@@ -115,54 +140,110 @@ func (c *Core) InsertChain(blocks types.Blocks) (int, error) {
-	for _, hashAndNumber := range hashNumberList[:threshold] {
-		block := c.GetBlockOrCandidateByHash(hashAndNumber.Hash)
-		if block != nil {
-			c.serviceFutureBlock(block)
+	for _, hashAndNumber := range hashNumberList {
+		header := c.GetHeaderOrCandidateByHash(hashAndNumber.Hash)
+		if header != nil {
+			var numberAndRetryCounter blockNumberAndRetryCounter
+			if value, exist := c.appendQueue.Peek(header.Hash()); exist {
+				numberAndRetryCounter = value.(blockNumberAndRetryCounter)
+				numberAndRetryCounter.retry += 1
+				if numberAndRetryCounter.retry > retryThreshold && numberAndRetryCounter.number+c_appendQueueRemoveThreshold < c.CurrentHeader().NumberU64() {
+					c.appendQueue.Remove(header.Hash())
+				} else {
+					c.appendQueue.Add(header.Hash(), numberAndRetryCounter)
+				}
+			}
+			parentHeader := c.GetHeader(header.ParentHash(), header.NumberU64()-1)
+			if parentHeader != nil {
+				// If parent header is dom, send a signal to dom to request for the block if it doesnt have it
+				_, parentHeaderOrder, err := c.sl.engine.CalcOrder(parentHeader)
+				if err != nil {
+					log.Warn("Error calculating the parent block order in serviceBlocks", "Hash", parentHeader.Hash(), "Number", parentHeader.NumberArray())
+				}
+				nodeCtx := common.NodeLocation.Context()
+				if parentHeaderOrder < nodeCtx && c.GetHeaderByHash(parentHeader.Hash()) == nil {
+					log.Info("Requesting the dom to get the block if it doesnt have and try to append", "Hash", parentHeader.Hash(), "Order", parentHeaderOrder)
+					// send a signal to the required dom to fetch the block if it doesnt have it, or its not in its appendqueue
+					go c.sl.domClient.RequestDomToAppendOrFetch(context.Background(), parentHeader.Hash(), parentHeaderOrder)
+				}
+				c.InsertChain([]*types.Block{types.NewBlockWithHeader(header)})
+			} else {
+				if !c.HasHeader(header.ParentHash(), header.NumberU64()-1) {
+					c.sl.missingParentFeed.Send(header.ParentHash())
+				}
+			}
 		} else {
 			log.Warn("Entry in the FH cache without being in the db: ", "Hash: ", hashAndNumber.Hash)
 		}
 	}
 }
 
-func (c *Core) serviceFutureBlock(block *types.Block) {
-	parentBlock := c.GetBlock(block.ParentHash(), block.NumberU64()-1)
-	if parentBlock != nil {
-		c.InsertChain([]*types.Block{block})
-	} else {
-		if !c.HasHeader(block.ParentHash(), block.NumberU64()-1) {
-			c.sl.missingParentFeed.Send(block.ParentHash())
+func (c *Core) RequestDomToAppendOrFetch(hash common.Hash, order int) {
+	// TODO: optimize to check if the block is in the appendqueue or already
+	// appended to reduce the network bandwidth utilization
+	nodeCtx := common.NodeLocation.Context()
+	if nodeCtx == common.PRIME_CTX {
+		// If prime all you can do it to ask for the block
+		_, exists := c.appendQueue.Get(hash)
+		if !exists {
+			log.Warn("Block sub asked doesnt exist in append queue, so request the peers for it", "Hash", hash, "Order", order)
+			c.sl.missingParentFeed.Send(hash)
```

Comment:
Just making sure, our peers will give us the block and not the block's parent, right?

Reply:
yes, it is, we have one feed that is called the missing parent feed, so reusing it here, can add a comment here to make sure there is no misunderstanding